04-vue-cli/03-UiDropdown1/components/UiDropdown.vue

```diff
@@ -1,31 +1,106 @@
 <template>
-  <div class="dropdown dropdown_opened">
-    <button type="button" class="dropdown__toggle dropdown__toggle_icon">
-      <UiIcon icon="tv" class="dropdown__icon" />
-      <span>Title</span>
+  <div class="dropdown" :class="{ dropdown_opened: isOpened }">
+    <button
+      type="button"
+      class="dropdown__toggle"
+      :class="{ dropdown__toggle_icon: hasIcon }"
+      @click="
+        () => {
+          isOpened = !isOpened;
+        }
+      "
+    >
+      <UiIcon v-if="value?.icon" :icon="value.icon" class="dropdown__icon" />
+      <span>{{ value?.text || title }}</span>
     </button>
 
-    <div class="dropdown__menu" role="listbox">
-      <button class="dropdown__item dropdown__item_icon" role="option" type="button">
-        <UiIcon icon="tv" class="dropdown__icon" />
-        Option 1
-      </button>
-      <button class="dropdown__item dropdown__item_icon" role="option" type="button">
-        <UiIcon icon="tv" class="dropdown__icon" />
-        Option 2
+    <div v-show="isOpened" class="dropdown__menu" role="listbox">
+      <button
+        v-for="option in options"
+        class="dropdown__item"
+        :class="{ dropdown__item_icon: hasIcon }"
+        :key="option.value"
+        @click="() => select(option.value)"
+        role="option"
+        type="button"
+      >
+        <UiIcon v-if="option.icon" :icon="option.icon" class="dropdown__icon" />
+        {{ option.text }}
       </button>
     </div>
+
+    <select v-show="false" :value="value?.value" :text="value?.text" @change="selectNative">
+      <option v-for="option in options" :value="option.value" :key="option.value">{{ option.text }}</option>
+    </select>
   </div>
 </template>
 
-<script>
+<script lang="ts">
+import { defineComponent, type PropType } from 'vue';
 import UiIcon from './UiIcon.vue';
 
-export default {
+type TItem = { value: string; text: string; icon?: string };
+type TData = {
+  isOpened: boolean;
+};
+
+export default defineComponent({
   name: 'UiDropdown',
 
   components: { UiIcon },
-};
+
+  props: {
+    options: {
+      type: Array as PropType<TItem[]>,
+      default: () => [],
+    },
+
+    modelValue: {
+      type: String,
+    },
+
+    title: {
+      type: String,
+      required: true,
+    },
+  },
+
+  emits: ['update:modelValue'],
+
+  data(): TData {
+    return {
+      isOpened: false,
+    };
+  },
+
+  computed: {
+    hasIcon(): boolean {
+      return this.options.some(({ icon }) => icon);
+    },
+
+    value: {
+      get(): TItem | undefined {
+        return this.options.find(({ value }) => value === this.modelValue);
+      },
+
+      set({ value }: TItem) {
+        this.$emit('update:modelValue', value);
+      },
+    },
+  },
+
+  methods: {
+    select(value: string) {
+      this.value = this.options.find((it) => it.value === value);
+      this.isOpened = false;
+    },
```

Comment:
On `select` here we first loop all elements to find an option with `it.value === value`. And then in the setter extract value from this element for emit. The same `value` we used in the search. We could have emitted this value without looping.

The problem here is that the setter is used for emitting. It is not very implicit. A developer usually doesn't expect that value assignment is an action like emitting. Computed with getter+setter is useful for proxying `v-model`, when we need to send something as a value to `v0-model` with custom setter.